File: src/compilers/support/generic/gpio.cpp
// Copyright [2025] Nathan Skipper
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

#include "gpio.h"

#include <cstdlib>
#include <fstream>

#ifdef __linux__
#include <filesystem>
#endif

GPIOClient::GPIOClient() : IOClient("GPIO") {}

void GPIOClient::connect()
{
#ifdef __linux__
    connected = true;
#else
    connected = false;
#endif
}

void GPIOClient::onMappingAdded(const IOMap &map)
{
#ifdef __linux__
    int startPin = -1;
    if (!resolveGlobalPin(map, startPin))
    {
        return;
    }
    const std::string direction = (map.direction == IOType::Output) ? "out" : "in";
    const int width = (map.width > 1) ? map.width : 1;
    for (int bit = 0; bit < width; ++bit)
    {
        const int pin = startPin + bit;
        if (!ensureExported(pin))
        {
            return;
        }
        if (!ensureDirection(pin, direction))
        {
            return;
        }
    }

    pinByRemote[map.remoteAddress] = startPin;
#else
    (void)map;
#endif
}

bool GPIOClient::readBit(const std::string &remote, int &result)
{
#ifdef __linux__
    int globalPin = -1;
    if (!resolveRemotePin(remote, globalPin))
    {
        return false;
    }
    return readGpioValue(globalPin, result);
#else
    (void)remote;
    result = 0;
    return false;
#endif
}

bool GPIOClient::writeBit(const std::string &remote, int value)
{
#ifdef __linux__
    int globalPin = -1;
    if (!resolveRemotePin(remote, globalPin))
    {
        return false;
    }
    return writeGpioValue(globalPin, value);
#else
    (void)remote;
    (void)value;
    return false;
#endif
}

bool GPIOClient::readByte(const std::string &remote, uint8_t &result)
{
#ifdef __linux__
    int startPin = -1;
    if (!resolveRemotePin(remote, startPin))
    {
        return false;
    }

    result = 0;
    for (int bit = 0; bit < 8; ++bit)
    {
        int pinValue = 0;
        if (!readGpioValue(startPin + bit, pinValue))
        {
            return false;
        }
        if (pinValue != 0)
        {
            result |= static_cast<uint8_t>(1U << bit);
        }
    }
    return true;
#else
    (void)remote;
    result = 0;
    return false;
#endif
}

bool GPIOClient::writeByte(const std::string &remote, uint8_t value)
{
#ifdef __linux__
    int startPin = -1;
    if (!resolveRemotePin(remote, startPin))
    {
        return false;
    }

    for (int bit = 0; bit < 8; ++bit)
    {
        const int pinValue = ((value >> bit) & 0x01U) != 0 ? 1 : 0;
        if (!writeGpioValue(startPin + bit, pinValue))
        {
            return false;
        }
    }
    return true;
#else
    (void)remote;
    (void)value;
    return false;
#endif
}

bool GPIOClient::readWord(const std::string &remote, uint16_t &result)
{
#ifdef __linux__
    int startPin = -1;
    if (!resolveRemotePin(remote, startPin))
    {
        return false;
    }

    result = 0;
    for (int bit = 0; bit < 16; ++bit)
    {
        int pinValue = 0;
        if (!readGpioValue(startPin + bit, pinValue))
        {
            return false;
        }
        if (pinValue != 0)
        {
            result |= static_cast<uint16_t>(1U << bit);
        }
    }
    return true;
#else
    (void)remote;
    result = 0;
    return false;
#endif
}

bool GPIOClient::writeWord(const std::string &remote, uint16_t value)
{
#ifdef __linux__
    int startPin = -1;
    if (!resolveRemotePin(remote, startPin))
    {
        return false;
    }

    for (int bit = 0; bit < 16; ++bit)
    {
        const int pinValue = ((value >> bit) & 0x01U) != 0 ? 1 : 0;
        if (!writeGpioValue(startPin + bit, pinValue))
        {
            return false;
        }
    }
    return true;
#else
    (void)remote;
    (void)value;
    return false;
#endif
}

bool GPIOClient::readDWord(const std::string &remote, uint32_t &result)
{
#ifdef __linux__
    int startPin = -1;
    if (!resolveRemotePin(remote, startPin))
    {
        return false;
    }

    result = 0;
    for (int bit = 0; bit < 32; ++bit)
    {
        int pinValue = 0;
        if (!readGpioValue(startPin + bit, pinValue))
        {
            return false;
        }
        if (pinValue != 0)
        {
            result |= (static_cast<uint32_t>(1) << bit);
        }
    }
    return true;
#else
    (void)remote;
    result = 0;
    return false;
#endif
}

bool GPIOClient::writeDWord(const std::string &remote, uint32_t value)
{
#ifdef __linux__
    int startPin = -1;
    if (!resolveRemotePin(remote, startPin))
    {
        return false;
    }

    for (int bit = 0; bit < 32; ++bit)
    {
        const int pinValue = ((value >> bit) & static_cast<uint32_t>(1)) != 0 ? 1 : 0;
        if (!writeGpioValue(startPin + bit, pinValue))
        {
            return false;
        }
    }
    return true;
#else
    (void)remote;
    (void)value;
    return false;
#endif
}

bool GPIOClient::readLWord(const std::string &remote, uint64_t &result)
{
#ifdef __linux__
    int startPin = -1;
    if (!resolveRemotePin(remote, startPin))
    {
        return false;
    }

    result = 0;
    for (int bit = 0; bit < 64; ++bit)
    {
        int pinValue = 0;
        if (!readGpioValue(startPin + bit, pinValue))
        {
            return false;
        }
        if (pinValue != 0)
        {
            result |= (static_cast<uint64_t>(1) << bit);
        }
    }
    return true;
#else
    (void)remote;
    result = 0;
    return false;
#endif
}

bool GPIOClient::writeLWord(const std::string &remote, uint64_t value)
{
#ifdef __linux__
    int startPin = -1;
    if (!resolveRemotePin(remote, startPin))
    {
        return false;
    }

    for (int bit = 0; bit < 64; ++bit)
    {
        const int pinValue = ((value >> bit) & static_cast<uint64_t>(1)) != 0 ? 1 : 0;
        if (!writeGpioValue(startPin + bit, pinValue))
        {
            return false;
        }
    }
    return true;
#else
    (void)remote;
    (void)value;
    return false;
#endif
}

#ifdef __linux__
std::string GPIOClient::normalizeChipName(const std::string &chip) const
{
    if (chip.empty())
    {
        return "gpiochip0";
    }

    const std::string lower = toLowerCase(chip);
    if (lower.rfind("gpiochip", 0) == 0)
    {
        return lower;
    }

    return "gpiochip" + lower;
}

bool GPIOClient::readIntFile(const std::string &path, int &value) const
{
    std::ifstream in(path);
    if (!in)
    {
        return false;
    }
    in >> value;
    return in.good() || in.eof();
}

bool GPIOClient::writeTextFile(const std::string &path, const std::string &value) const
{
    std::ofstream out(path);
    if (!out)
    {
        return false;
    }
    out << value;
    return out.good();
}

bool GPIOClient::ensureExported(int globalPin) const
{
    const std::string gpioPath = "/sys/class/gpio/gpio" + std::to_string(globalPin);
    if (std::filesystem::exists(gpioPath))
    {
        return true;
    }

    if (!writeTextFile("/sys/class/gpio/export", std::to_string(globalPin)))
    {
        return std::filesystem::exists(gpioPath);
    }

    return true;
}

bool GPIOClient::ensureDirection(int globalPin, const std::string &direction) const
{
    return writeTextFile("/sys/class/gpio/gpio" + std::to_string(globalPin) + "/direction", direction);
}

bool GPIOClient::resolveGlobalPin(const IOMap &map, int &globalPin) const
{
    const std::string chipName = normalizeChipName(map.moduleID);
    const std::string chipBasePath = "/sys/class/gpio/" + chipName + "/base";

    char *endPtr = nullptr;
    const long offset = std::strtol(map.remoteAddress.c_str(), &endPtr, 10);
    if (endPtr == map.remoteAddress.c_str() || offset < 0)
    {
        return false;
    }

    int base = 0;
    if (!readIntFile(chipBasePath, base))
    {
        // Fallback to absolute pin numbering if gpiochip base is unavailable.
        globalPin = static_cast<int>(offset);
        return true;
    }

    globalPin = base + static_cast<int>(offset);
    return true;
}

bool GPIOClient::resolveRemotePin(const std::string &remote, int &globalPin) const
{
    const auto it = pinByRemote.find(remote);
    if (it != pinByRemote.end())
    {
        globalPin = it->second;
        return true;
    }

    IOMap temp;
    temp.moduleID = moduleID;
    temp.remoteAddress = remote;
    return resolveGlobalPin(temp, globalPin);
}

bool GPIOClient::readGpioValue(int globalPin, int &value) const
{
    std::ifstream in("/sys/class/gpio/gpio" + std::to_string(globalPin) + "/value");
    if (!in)
    {
        return false;
    }

    char c = '0';
    in >> c;
    if (!in)
    {
        return false;
    }

    value = (c == '0') ? 0 : 1;
    return true;
}

bool GPIOClient::writeGpioValue(int globalPin, int value) const
{
    return writeTextFile("/sys/class/gpio/gpio" + std::to_string(globalPin) + "/value", (value == 0) ? "0" : "1");
}
#endif

File: src/compilers/support/nodejs/gpio.js
// Copyright [2025] Nathan Skipper
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

import fs from "fs";
import { IOClient } from "./IOClient.js";

function isLinux() {
  return process.platform === "linux";
}

function normalizeChip(chip) {
  if (!chip || chip.length === 0) {
    return "gpiochip0";
  }

  const lower = String(chip).toLowerCase();
  if (lower.startsWith("gpiochip")) {
    return lower;
  }

  return `gpiochip${lower}`;
}

function parseIntSafe(value) {
  const parsed = Number.parseInt(value, 10);
  return Number.isNaN(parsed) ? null : parsed;
}

function readIntFile(path) {
  try {
    const raw = fs.readFileSync(path, "utf8").trim();
    const parsed = parseIntSafe(raw);
    return parsed === null ? null : parsed;
  } catch {
    return null;
  }
}

function writeTextFile(path, value) {
  try {
    fs.writeFileSync(path, value, "utf8");
    return true;
  } catch {
    return false;
  }
}

export class GPIOClient extends IOClient {
  constructor() {
    super("GPIO");
    this.pinByRemote = new Map();
  }

  connect() {
    this.connected = isLinux();
  }

  onMappingAdded(map) {
    if (!isLinux()) {
      return;
    }

    const startPin = this.resolveGlobalPin(map.moduleID, map.remoteAddress);
    if (startPin === null) {
      return;
    }

    const direction = map.direction === "Output" ? "out" : "in";
    const width = map.width > 1 ? map.width : 1;
    for (let bit = 0; bit < width; bit += 1) {
      const pin = startPin + bit;
      if (!this.ensureExported(pin)) {
        return;
      }
      if (!writeTextFile(`/sys/class/gpio/gpio${pin}/active_low`, "1")) {
        return;
      }
      if (!writeTextFile(`/sys/class/gpio/gpio${pin}/direction`, direction)) {
        return;
      }
    }

    this.pinByRemote.set(map.remoteAddress, startPin);
  }

  resolveGlobalPin(moduleID, remoteAddress) {
    const offset = parseIntSafe(remoteAddress);
    if (offset === null || offset < 0) {
      return null;
    }

    const chip = normalizeChip(moduleID);
    const base = readIntFile(`/sys/class/gpio/${chip}/base`);
    if (base === null) {
      // Fallback to absolute pin addressing when gpiochip metadata is unavailable.
      return offset;
    }

    return base + offset;
  }

  ensureExported(globalPin) {
    const gpioPath = `/sys/class/gpio/gpio${globalPin}`;
    if (fs.existsSync(gpioPath)) {
      return true;
    }

    if (!writeTextFile("/sys/class/gpio/export", String(globalPin))) {
      return fs.existsSync(gpioPath);
    }

    return true;
  }

  resolveRemotePin(remote) {
    if (this.pinByRemote.has(remote)) {
      return this.pinByRemote.get(remote);
    }

    const map = this.mappings.find(m => m.remoteAddress === remote);
    if (!map) {
      return null;
    }

    return this.resolveGlobalPin(map.moduleID || this.moduleID, remote);
  }

  readValue(remote, callback) {
    if (!isLinux()) {
      return;
    }

    const globalPin = this.resolveRemotePin(remote);
    if (globalPin === null) {
      return;
    }

    try {
      const raw = fs.readFileSync(`/sys/class/gpio/gpio${globalPin}/value`, "utf8").trim();
      callback(raw === "0" ? 0 : 1);
    } catch {
      // Intentionally ignored: GPIO reads are best-effort.
    }
  }

  writeValue(remote, value) {
    if (!isLinux()) {
      return;
    }

    const globalPin = this.resolveRemotePin(remote);
    if (globalPin === null) {
      return;
    }

    writeTextFile(`/sys/class/gpio/gpio${globalPin}/value`, value === 0 ? "0" : "1");
  }

  readBit(remote, callback) {
    this.readValue(remote, callback);
  }

  writeBit(remote, value) {
    this.writeValue(remote, value ? 1 : 0);
  }

  readWidthValue(remote, width, callback) {
    if (!isLinux()) {
      return;
    }

    const startPin = this.resolveRemotePin(remote);
    if (startPin === null) {
      return;
    }

    let value = 0n;
    for (let bit = 0; bit < width; bit += 1) {
      try {
        const raw = fs.readFileSync(`/sys/class/gpio/gpio${startPin + bit}/value`, "utf8").trim();
        if (raw !== "0") {
          value |= (1n << BigInt(bit));
        }
      } catch {
        return;
      }
    }

    callback(value);
  }

  writeWidthValue(remote, width, value) {
    if (!isLinux()) {
      return;
    }

    const startPin = this.resolveRemotePin(remote);
    if (startPin === null) {
      return;
    }

    const normalized = typeof value === "bigint" ? value : BigInt(value >>> 0);
    for (let bit = 0; bit < width; bit += 1) {
      const bitValue = ((normalized >> BigInt(bit)) & 1n) === 1n ? "1" : "0";
      if (!writeTextFile(`/sys/class/gpio/gpio${startPin + bit}/value`, bitValue)) {
        return;
      }
    }
  }

  readByte(remote, callback) {
    this.readWidthValue(remote, 8, v => callback(Number(v & 0xFFn)));
  }

  writeByte(remote, value) {
    this.writeWidthValue(remote, 8, BigInt(value & 0xFF));
  }

  readWord(remote, callback) {
    this.readWidthValue(remote, 16, v => callback(Number(v & 0xFFFFn)));
  }

  writeWord(remote, value) {
    this.writeWidthValue(remote, 16, BigInt(value & 0xFFFF));
  }

  readDWord(remote, callback) {
    this.readWidthValue(remote, 32, v => callback(Number(v & 0xFFFFFFFFn)));
  }

  writeDWord(remote, value) {
    this.writeWidthValue(remote, 32, BigInt(value >>> 0));
  }

  readLWord(remote, callback) {
    this.readWidthValue(remote, 64, callback);
  }

  writeLWord(remote, value) {
    this.writeWidthValue(remote, 64, value);
  }
}
